### Import libraries

In [2]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline

import warnings
warnings.filterwarnings('ignore')

In [3]:
from sklearn.datasets import load_breast_cancer

In [5]:
# load data
data = load_breast_cancer()['data']
cols = load_breast_cancer()['feature_names']

df = pd.DataFrame(data=data,columns=cols)
df['Target'] = load_breast_cancer()['target']

In [6]:
df.head()

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0        17.99         10.38          122.80     1001.0          0.11840   
1        20.57         17.77          132.90     1326.0          0.08474   
2        19.69         21.25          130.00     1203.0          0.10960   
3        11.42         20.38           77.58      386.1          0.14250   
4        20.29         14.34          135.10     1297.0          0.10030   

   mean compactness  mean concavity  mean concave points  mean symmetry  \
0           0.27760          0.3001              0.14710         0.2419   
1           0.07864          0.0869              0.07017         0.1812   
2           0.15990          0.1974              0.12790         0.2069   
3           0.28390          0.2414              0.10520         0.2597   
4           0.13280          0.1980              0.10430         0.1809   

   mean fractal dimension  ...  worst texture  worst perimeter  worst area  \
0                 0.07871  ...          17.33           184.60      2019.0   
1                 0.05667  ...          23.41           158.80      1956.0   
2                 0.05999  ...          25.53           152.50      1709.0   
3                 0.09744  ...          26.50            98.87       567.7   
4                 0.05883  ...          16.67           152.20      1575.0   

   worst smoothness  worst compactness  worst concavity  worst concave points  \
0            0.1622             0.6656           0.7119                0.2654   
1            0.1238             0.1866           0.2416                0.1860   
2            0.1444             0.4245           0.4504                0.2430   
3            0.2098             0.8663           0.6869                0.2575   
4            0.1374             0.2050           0.4000                0.1625   

   worst symmetry  worst fractal dimension  Target  
0          0.4601                  0.11890       0  
1          0.2750                  0.08902       0  
2          0.3613                  0.08758       0  
3          0.6638                  0.17300       0  
4          0.2364                  0.07678       0  

[5 rows x 31 columns]

In [7]:
df.shape

(569, 31)

In [8]:
# check missing data
df.isna().sum()

mean radius                0
mean texture               0
mean perimeter             0
mean area                  0
mean smoothness            0
mean compactness           0
mean concavity             0
mean concave points        0
mean symmetry              0
mean fractal dimension     0
radius error               0
texture error              0
perimeter error            0
area error                 0
smoothness error           0
compactness error          0
concavity error            0
concave points error       0
symmetry error             0
fractal dimension error    0
worst radius               0
worst texture              0
worst perimeter            0
worst area                 0
worst smoothness           0
worst compactness          0
worst concavity            0
worst concave points       0
worst symmetry             0
worst fractal dimension    0
Target                     0
dtype: int64

##### The dataset has 30 features and 1 target column

30 is very huge as far as dimensionality is concerned. Unnecessary features can act as NOISE

In [10]:
# descriptive statistics

df.describe().T

count        mean         std         min  \
mean radius              569.0   14.127292    3.524049    6.981000   
mean texture             569.0   19.289649    4.301036    9.710000   
mean perimeter           569.0   91.969033   24.298981   43.790000   
mean area                569.0  654.889104  351.914129  143.500000   
mean smoothness          569.0    0.096360    0.014064    0.052630   
mean compactness         569.0    0.104341    0.052813    0.019380   
mean concavity           569.0    0.088799    0.079720    0.000000   
mean concave points      569.0    0.048919    0.038803    0.000000   
mean symmetry            569.0    0.181162    0.027414    0.106000   
mean fractal dimension   569.0    0.062798    0.007060    0.049960   
radius error             569.0    0.405172    0.277313    0.111500   
texture error            569.0    1.216853    0.551648    0.360200   
perimeter error          569.0    2.866059    2.021855    0.757000   
area error               569.0   40.337079   45.491006    6.802000   
smoothness error         569.0    0.007041    0.003003    0.001713   
compactness error        569.0    0.025478    0.017908    0.002252   
concavity error          569.0    0.031894    0.030186    0.000000   
concave points error     569.0    0.011796    0.006170    0.000000   
symmetry error           569.0    0.020542    0.008266    0.007882   
fractal dimension error  569.0    0.003795    0.002646    0.000895   
worst radius             569.0   16.269190    4.833242    7.930000   
worst texture            569.0   25.677223    6.146258   12.020000   
worst perimeter          569.0  107.261213   33.602542   50.410000   
worst area               569.0  880.583128  569.356993  185.200000   
worst smoothness         569.0    0.132369    0.022832    0.071170   
worst compactness        569.0    0.254265    0.157336    0.027290   
worst concavity          569.0    0.272188    0.208624    0.000000   
worst concave points     569.0    0.114606    0.065732    0.000000   
worst symmetry           569.0    0.290076    0.061867    0.156500   
worst fractal dimension  569.0    0.083946    0.018061    0.055040   
Target                   569.0    0.627417    0.483918    0.000000   

                                25%         50%          75%         max  
mean radius               11.700000   13.370000    15.780000    28.11000  
mean texture              16.170000   18.840000    21.800000    39.28000  
mean perimeter            75.170000   86.240000   104.100000   188.50000  
mean area                420.300000  551.100000   782.700000  2501.00000  
mean smoothness            0.086370    0.095870     0.105300     0.16340  
mean compactness           0.064920    0.092630     0.130400     0.34540  
mean concavity             0.029560    0.061540     0.130700     0.42680  
mean concave points        0.020310    0.033500     0.074000     0.20120  
mean symmetry              0.161900    0.179200     0.195700     0.30400  
mean fractal dimension     0.057700    0.061540     0.066120     0.09744  
radius error               0.232400    0.324200     0.478900     2.87300  
texture error              0.833900    1.108000     1.474000     4.88500  
perimeter error            1.606000    2.287000     3.357000    21.98000  
area error                17.850000   24.530000    45.190000   542.20000  
smoothness error           0.005169    0.006380     0.008146     0.03113  
compactness error          0.013080    0.020450     0.032450     0.13540  
concavity error            0.015090    0.025890     0.042050     0.39600  
concave points error       0.007638    0.010930     0.014710     0.05279  
symmetry error             0.015160    0.018730     0.023480     0.07895  
fractal dimension error    0.002248    0.003187     0.004558     0.02984  
worst radius              13.010000   14.970000    18.790000    36.04000  
worst texture             21.080000   25.410000    29.720000    49.54000  
worst perimeter           84.110000   97.660000   125.40000

#### Lets set feature and target set

In [11]:
# X = feature
# Y= target

X = df.drop('Target',axis=1)
Y =df['Target']

In [12]:
X.shape

(569, 30)

In [13]:
Y.shape

(569,)

### Train Test Split

In [14]:
from sklearn.model_selection import train_test_split

In [15]:
x_train,x_test,y_train,y_test = train_test_split(X,Y,test_size=0.3,random_state=42)

#### Standardize the dataset

In [16]:
from sklearn.preprocessing import StandardScaler

In [17]:
scaler = StandardScaler()

x_train_scaled = scaler.fit_transform(x_train)
x_test_scaled = scaler.fit_transform(x_test)

#### Define RandomForest Classifier to be used by Boruta

In [18]:
from sklearn.ensemble import RandomForestClassifier

In [19]:
model = RandomForestClassifier() # For Boruta

### Boruta 

In [20]:
from boruta import BorutaPy

In [21]:
# define Boruta feature selection method
feature_selector = BorutaPy(model,n_estimators='auto',verbose=2,random_state=1,max_iter=100)

In [22]:
feature_selector.fit(np.array(x_train),np.array(y_train))

Iteration: 	1 / 100
Confirmed: 	0
Tentative: 	30
Rejected: 	0
Iteration: 	2 / 100
Confirmed: 	0
Tentative: 	30
Rejected: 	0
Iteration: 	3 / 100
Confirmed: 	0
Tentative: 	30
Rejected: 	0
Iteration: 	4 / 100
Confirmed: 	0
Tentative: 	30
Rejected: 	0
Iteration: 	5 / 100
Confirmed: 	0
Tentative: 	30
Rejected: 	0
Iteration: 	6 / 100
Confirmed: 	0
Tentative: 	30
Rejected: 	0
Iteration: 	7 / 100
Confirmed: 	0
Tentative: 	30
Rejected: 	0
Iteration: 	8 / 100
Confirmed: 	19
Tentative: 	11
Rejected: 	0
Iteration: 	9 / 100
Confirmed: 	19
Tentative: 	11
Rejected: 	0
Iteration: 	10 / 100
Confirmed: 	19
Tentative: 	10
Rejected: 	1
Iteration: 	11 / 100
Confirmed: 	19
Tentative: 	10
Rejected: 	1
Iteration: 	12 / 100
Confirmed: 	19
Tentative: 	8
Rejected: 	3
Iteration: 	13 / 100
Confirmed: 	19
Tentative: 	8
Rejected: 	3
Iteration: 	14 / 100
Confirmed: 	19
Tentative: 	8
Rejected: 	3
Iteration: 	15 / 100
Confirmed: 	19
Tentative: 	8
Rejected: 	3
Iteration: 	16 / 100
Confirmed: 	19
Tentative: 	8
Rejected: 

BorutaPy(estimator=RandomForestClassifier(n_estimators=69,
                                          random_state=RandomState(MT19937) at 0x1B8B3014B40),
         n_estimators='auto',
         random_state=RandomState(MT19937) at 0x1B8B3014B40, verbose=2)

Print the decisions made by the feature_selector

In [23]:
print('This array has on wheter to keep a feature or not!\n')
print(feature_selector.support_)

This array has on wheter to keep a feature or not!

[ True  True  True  True  True  True  True  True False False  True False
  True  True False False  True False False False  True  True  True  True
  True  True  True  True  True  True]


In [24]:
np.sum(feature_selector.support_)

22

In [25]:
# print the array has feature ranking

print(feature_selector.ranking_)

[1 1 1 1 1 1 1 1 5 3 1 8 1 1 7 5 1 2 5 8 1 1 1 1 1 1 1 1 1 1]


In [26]:
x_filtered_train = feature_selector.transform(np.array(x_train))
x_filtered_test = feature_selector.transform(np.array(x_test))

### Lets Compare Performances


In [27]:
# model without BorutaPy
model1 = RandomForestClassifier()

In [28]:
#train the model

model1.fit(x_train_scaled,y_train)

RandomForestClassifier()

In [29]:
## Test the model1

y_pred = model1.predict(x_test_scaled)

In [32]:
from sklearn.metrics import accuracy_score, classification_report

In [31]:
print(accuracy_score(y_test,y_pred))

0.9707602339181286


In [33]:
print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

           0       0.98      0.94      0.96        63
           1       0.96      0.99      0.98       108

    accuracy                           0.97       171
   macro avg       0.97      0.96      0.97       171
weighted avg       0.97      0.97      0.97       171



In [39]:
pd.crosstab(y_test,y_pred)

col_0    0    1
Target         
0       59    4
1        1  107

#### Lets check the model with BorutaPy

In [34]:
#train the model
model.fit(x_filtered_train,y_train)

RandomForestClassifier(n_estimators=69,
                       random_state=RandomState(MT19937) at 0x1B8B3014B40)

In [35]:
# test the model
yb_pred = model.predict(x_filtered_test)

In [36]:
print(classification_report(y_test,yb_pred))

              precision    recall  f1-score   support

           0       0.97      0.92      0.94        63
           1       0.95      0.98      0.97       108

    accuracy                           0.96       171
   macro avg       0.96      0.95      0.96       171
weighted avg       0.96      0.96      0.96       171



In [37]:
print(accuracy_score(y_test,yb_pred))

0.9590643274853801


In [38]:
pd.crosstab(y_test,yb_pred)

col_0    0    1
Target         
0       58    5
1        2  106

The model with BorutaPy gave an accuracy score of ~96%.
And the model without BorutaPy gave an accuracy score of 97%